modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java

```diff
@@ -231,6 +234,59 @@ public static BigDecimal sround(BigDecimal b0, int b1) {
         return roundedValue.setScale(originalScale, RoundingMode.UNNECESSARY);
     }
 
+    /** Returns {@link Integer#MAX_VALUE} if input is greater than {@link Integer#MAX_VALUE},
+     * otherwise return integer input number representation. */
+    private static int normalizeRegardingInt(BigDecimal num) {
+        int res;
+
+        if (num.compareTo(MAX_INT) > 0) {
+            res = Integer.MAX_VALUE;
+        } else if (num.compareTo(MIN_INT) < 0) {
+            res = Integer.MIN_VALUE;
+        } else {
+            res = num.intValue();
+        }
+
+        return res;
+    }
+
+    /** SQL SUBSTRING(string FROM ...) function. */
+    public static String substring(String c, int s) {
+        if (s < 0) {
```

Comment:
redundant check - original function alredy has the optimizations, even wider then this one

Reply:
Unfortunately this check is necessary due to calcite bug [1]

[1] https://issues.apache.org/jira/browse/CALCITE-6211

append additional tests.

Reply:
in this case it looks should be  `s<=1`

Reply:
Can we just call `int j = IgniteMath.subtractExact(s, 1)` to add an overflow check.
and then call `IgniteSqlMethid::substring(c, s, j)` ?